src/cargo/ops/cargo_update.rs

```diff
@@ -207,6 +214,251 @@ pub fn print_lockfile_changes(
+) -> CargoResult<Dependency> {
+    let name = dependency.package_name();
+    let renamed_to = dependency.name_in_toml();
+
+    if name != renamed_to {
+        trace!(
+            "skipping dependency renamed from `{}` to `{}`",
+            name,
+            renamed_to
+        );
+        return Ok(dependency);
+    }
+
+    if !opts.to_update.is_empty() && !opts.to_update.contains(&name.to_string()) {
+        trace!("skipping dependency `{}` not selected for upgrading", name);
+        return Ok(dependency);
+    }
+
+    let version_req = dependency.version_req();
+
+    let OptVersionReq::Req(current) = version_req else {
+        trace!(
+            "skipping dependency `{}` without a simple version requirement: {}",
+            name,
+            version_req
+        );
+        return Ok(dependency);
+    };
+
+    let [comparator] = &current.comparators[..] else {
+        trace!(
+            "skipping dependency `{}` with multiple version comparators: {:?}",
+            name,
+            &current.comparators
+        );
+        return Ok(dependency);
+    };
+
+    if comparator.op != Op::Caret {
+        trace!("skipping non-caret dependency `{}`: {}", name, comparator);
+        return Ok(dependency);
+    }
+
+    let query =
+        crate::core::dependency::Dependency::parse(name, None, dependency.source_id().clone())?;
+
+    let possibilities = {
+        loop {
+            match registry.query_vec(&query, QueryKind::Exact) {
+                std::task::Poll::Ready(res) => {
+                    break res?;
+                }
+                std::task::Poll::Pending => registry.block_until_ready()?,
+            }
+        }
+    };
+
+    let latest = if !possibilities.is_empty() {
+        possibilities
+            .iter()
+            .map(|s| s.as_summary())
+            .map(|s| s.version())
+            .filter(|v| !v.is_prerelease())
+            .max()
+    } else {
+        None
+    };
+
+    let Some(latest) = latest else {
+        trace!(
+            "skipping dependency `{}` without any published versions",
+            name
+        );
+        return Ok(dependency);
+    };
+
+    if current.matches(&latest) {
+        trace!(
+            "skipping dependency `{}` without a breaking update available",
+            name
+        );
+        return Ok(dependency);
+    }
+
+    let Some(new_req_string) = upgrade_requirement(&current.to_string(), latest)? else {
+        trace!(
+            "skipping dependency `{}` because the version requirement didn't change",
+            name
+        );
+        return Ok(dependency);
+    };
+
+    let upgrade_message = format!("{} {} -> {}", name, current, new_req_string);
+    trace!(upgrade_message);
+
+    if upgrade_messages.insert(upgrade_message.clone()) {
+        opts.gctx
+            .shell()
+            .status_with_color("Upgrading", &upgrade_message, &style::GOOD)?;
+    }
+
+    upgrades.insert(name.to_string(), latest.clone());
```

Comment:
Since you are using the package name, rather than the dependency name, we should likely always `insert` the same thing here (assuming its the same registry).

There is an optimization that could be done to skip some of the above steps if an entry already exists in `upgrades` but likely not important.